File: HOT100/9.2/207.课程表.py
#
# @lc app=leetcode.cn id=207 lang=python3
#
# [207] 课程表
#
from mytools import *
# @lc code=start
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        g = [[] for _ in range(numCourses)]
        ind = [0] * numCourses
        for u, v in prerequisites:
            g[v].append(u)
            ind[u] += 1
        q = deque()
        for i in range(numCourses):
            if ind[i] == 0: q.append(i)
        cnt = 0
        while q:
            u = q.popleft()
            cnt += 1
            for v in g[u]:
                ind[v] -= 1
                if ind[v] == 0: q.append(v)
        return cnt == numCourses
# @lc code=end

File: HOT100/HOT100p1/9.图论/207.课程表.cpp
/*
 * @lc app=leetcode.cn id=207 lang=cpp
 *
 * [207] 课程表
 */
#include "tools.h"
// @lc code=start
class Solution
{
public:
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {
        vector<vector<int>> g(numCourses);
        vector<int> ind(numCourses);
        for (auto& p : prerequisites)
        {
            int u = p[0], v = p[1];
            g[v].push_back(u);
            ind[u] ++;
        }
        int cnt = 0;
        queue<int> q;
        for (int i = 0; i < numCourses; i ++ )
        {
            if (ind[i] == 0) q.push(i);
        }
        while (q.size())
        {
            int u = q.front(); q.pop();
            cnt ++;
            for (auto& v : g[u])
            {
                if (-- ind[v] == 0) q.push(v);
            }
        }
        return cnt == numCourses;
    }
};
// @lc code=end

class Solution
{
public:
    bool canFinish(int numCourses, vector<vector<int>> &prerequisites)
    {

        vector<unordered_set<int>> g(numCourses);
        for (auto &p : prerequisites)
        {
            int u = p[0], v = p[1];
            g[u].insert(v);
        }
        int ok = 0;
        int prevok = -1;
        vector<bool> rm(numCourses, 0);
        while (ok < numCourses)
        {
            if (prevok == ok)
                break;
            prevok = ok;
            for (int i = 0; i < numCourses; i++)
            {
                if (g[i].empty() && !rm[i])
                {
                    ok++;
                    rm[i] = 1;
                    for (int j = 0; j < numCourses; j++)
                    {
                        if (g[j].count(i))
                            g[j].erase(i);
                    }
                }
            }
        }
        return ok == numCourses;
    }
};